docs/designs/restore-and-gc.md

```diff
@@ -0,0 +1,160 @@
+
+Config(ocs/designs/config.md):
+
+``` yml
+dependencies:
+  docfx-code: https://github.com/dotnet/docfx#master
+```
+
+Content:
+
+```md
+Example: [!code-csharp[Main](docfx-code/program.cs)]
+```
+
+The restored dependency repositories will be stored at `%DOCFX_APPDATA_PATH%/git/{url-short-path}/{worktree-name}` organized by [git work tree](https://git-scm.com/docs/git-worktree) like:
+
+```text
+`%DOCFX_APPDATA_PATH%`
+    | - git
+        | - github.com+Microsoft+Luis-Samples+5e455995
+            | - .git
+            | - master-fd6968dec9a9a39aec1845232466fe35fca520da
+            | - live-a82cb2a51f6925b2fa87275218b01678c5aadfe1
+```
+
+- `{url-short-path}` is calculated from `git remote` url
+- `{worktree-name}` is combined by `branch name` and `HEAD commit id`
+
+## Dependency lock
+
+There are some cases you want to keep using the **same version** of dependency files/repository between different `build`s.
+
+For example, if there are multiple contributors for your content, and there is a CI process to check everyone's change before merging their PRs, you truly want everyone including CI to use the same versions of the dependency files/repositories for builds, to provide valid checking result and reduce influences of different versions of dependencies.
+
+Docfx uses `.lock.json`. Whenever you run docfx `restore`, docfx generates or updates your `dependency lock`, which will look something like this:
+
+```json
+{
+    "git":{
+        "https://github.com/docfx-code-sample#master":{
+            "version": "fd6968dec9a9a39aec1845232466fe35fca520da"
+        },
+        "https://github.com/docfx-code-sample#live":{
+            "version": "a82cb2a51f6925b2fa87275218b01678c5aadfe1",
+            "git":{
+                "https://github.com/azure-code-sample#live":{
+                    "version": "c9a9a39aec184523246678c5aadfea82cb2"
+                }
+            },
+            "downloads":{
+                "https:///some-blob-service/build-history.json":{
+                    "version": "66458170ba048c7fc2de4+%18b265bc4e9a6b4fdff8b5210caf9e357%22"
+                }
+            }
+        }
+    },
+    "downloads":{
+        "https:///some-blob-service/github-user-cache.json":{
+            "version": "18b265bc4e9a6bf4cb2d+%22f02fde5b4f3817266458170ba048c7fc2de46287%22"
+        }
+    }
+}
+```
+
+- `git` is the version mappings of dependency repositories
+- `downloads` is the version mappings of dependency files
+- `version` is `commit id` of the dependency repository  or `content sha1` of the dependency file
+
+This file describes an `exact`, and more importantly `reproducible` `dependencies` trees. Once it's present, any future `restore` will base its work off this file, instead of recalculating`.
+
+The presence of the `dependency lock` changes the `restore` behavior:
+
+- The dependency files/repository described by this `dependency lock` is reproduced, which means reproducing the structure/version described in the file.
```

Comment:
add a section that lock file generation process `docfx.yml + .lock.json --> .lock.json` should be _idempotent_, and that content inside `.lock.json` is always be sorted.